src/group/managed_group/mod.rs

```diff
@@ -929,85 +950,67 @@ impl<'a> ManagedGroup<'a> {
         for proposal_or_ref in proposals {
             match proposal_or_ref {
                 ProposalOrRef::Proposal(proposal) => {
-                    self.send_proposal_event(proposal, sender, indexed_members);
+                    events.push(self.prepare_proposal_event(proposal, sender, indexed_members));
                 }
                 ProposalOrRef::Reference(proposal_reference) => {
                     if let Some(queued_proposal) = pending_proposals_queue.get(proposal_reference) {
-                        self.send_proposal_event(
+                        events.push(self.prepare_proposal_event(
                             queued_proposal.proposal(),
                             queued_proposal.sender().to_leaf_index(),
                             indexed_members,
-                        );
+                        ));
                     }
                 }
             }
         }
+        events
     }
 
-    /// Send out the corresponding events for the pending proposal list.
-    fn send_proposal_event(
+    /// Prepare the corresponding events for the pending proposal list.
+    fn prepare_proposal_event(
         &self,
         proposal: &Proposal,
         sender: LeafIndex,
         indexed_members: &HashMap<LeafIndex, Credential>,
-    ) {
+    ) -> GroupEvent {
         let sender_credential = &indexed_members[&sender];
         match proposal {
             // Add proposals
-            Proposal::Add(add_proposal) => {
-                if let Some(member_added) = self.managed_group_config.callbacks.member_added {
-                    member_added(
-                        &self,
-                        &self.aad,
-                        sender_credential,
-                        add_proposal.key_package.credential(),
-                    )
-                }
-            }
+            Proposal::Add(add_proposal) => GroupEvent::MemberAdded(MemberAddedEvent::new(
+                self.aad.to_vec(),
+                sender_credential.clone(),
+                add_proposal.key_package.credential().clone(),
+            )),
             // Update proposals
             Proposal::Update(update_proposal) => {
-                if let Some(member_updated) = self.managed_group_config.callbacks.member_updated {
-                    member_updated(&self, &self.aad, update_proposal.key_package.credential())
-                }
+                GroupEvent::MemberUpdated(MemberUpdatedEvent::new(
+                    self.aad.to_vec(),
+                    update_proposal.key_package.credential().clone(),
+                ))
             }
             // Remove proposals
             Proposal::Remove(remove_proposal) => {
                 let removal = Removal::new(
-                    self.credential_bundle.credential(),
-                    sender_credential,
-                    &indexed_members[&LeafIndex::from(remove_proposal.removed)],
+                    self.credential_bundle.credential().clone(),
```

Comment:
There are some additional clones here and a bit below. I assume this was previously done in the callbacks?

Reply:
As mentioned out-of-band: We have to clone all the data since we can no longer use references to data that is stored inside the group. Now that we return the events directly, we have to assume that they are long-lived and will outlive the ephemeral group state. While this creates some data duplication, I don't see any way around it conceptually.